source/CODE.md

```diff
@@ -1,33 +1,377 @@
+  - Memory: `raw_ptr.rs`, `simple_pptr.rs`, `layout.rs`
+  - Utilities: `bytes.rs`, `string.rs`, `slice.rs`, `hash_map.rs`, `hash_set.rs`
+  - Specifications for Rust's standard library: `std_specs/` directory
+- **state_machines_macros**: Macros for [concurrency-related verification](https://verus-lang.github.io/verus/state_machines/)
+- **rust_verify_test_macros**: Test utilities
+- **cargo-verus**: [Cargo integration](./docs/CARGO-VERUS.md)
+
+### Vendored dependencies
+
+The `dependencies/` directory contains forked versions of external crates:
+
+- **syn**: Verus extends Rust syntax with verification constructs (e.g., `spec`,
+  `proof`, `requires`, `ensures`, `Ghost<T>`). The forked `syn` crate adds
+  parsing support for these extensions so that `builtin_macros` can parse Verus
+  code inside `verus! { }` blocks.
+- **prettyplease**: A Rust code formatter used for generating readable output.
+  The fork includes modifications to handle Verus-specific syntax when
+  pretty-printing transformed code.
+
+These forks are necessary because Verus syntax is not valid standard Rust, so
+the standard `syn` parser would reject it. By extending `syn`, Verus can parse
+its custom syntax, transform it appropriately, and emit valid Rust for `rustc`.
+
+## Detailed Pipeline Stages
+
+
+### Stage 1: Rustc Integration
+
+Verus runs multiple rustc invocations (see detailed comment in `driver.rs`):
+
+1. **Verification pass**: Full verification with all ghost code present
+2. **Compilation pass** (when `--compile` is used): Compiles with ghost code erased for executable output
+
+The `VerifierCallbacksEraseMacro` implements `rustc_driver::Callbacks` to hook into rustc:
+- `config()`: Sets up query provider overrides
+- `after_expansion()`: Called after macro expansion, triggers VIR construction and verification
+
+### Stage 2: HIR to VIR-AST
+
+**Entry**: `rust_to_vir::crate_to_vir()` called from `Verifier::construct_vir_crate()`
+
+The conversion iterates through HIR items and dispatches to specialized handlers:
+
+- Functions: `rust_to_vir_func::check_item_fn()`
+- Structs: `rust_to_vir_adts::check_item_struct()`
+- Enums: `rust_to_vir_adts::check_item_enum()`
+- Traits/impls, type aliases, constants, statics
+
+**Type conversion**: `rust_to_vir_base::mid_ty_to_vir()` converts rustc `Ty<'tcx>` to VIR `Typ`
+
+**Key operations:**
+- Extract Verus attributes (`#[verifier(...)]`)
+- Convert Rust types to VIR types
+- Parse specifications (requires, ensures, invariants)
+- Handle mode annotations (spec, proof, exec)
+
+### Stage 3: VIR Simplification
+
+After VIR construction, several passes run:
+
+1. `ast_simplify::simplify_krate()`: Optimizations and normalizations
+2. `modes::check_crate()`: Validate Spec/Proof/Exec mode usage
+3. `well_formed::check_crate()`: Structural validation
+4. `prune::prune_krate_for_module_or_krate()`: Remove unused definitions
+5. `traits::merge_external_traits()`: Resolve trait implementations
```

Comment:
These are out of order. well_formed comes before modes and simplify I think is last. Not sure about the others.